TST, DOC: Added Broadcasting Tests in test_random.py

gfyoung opened the pull request:
Follow-up from #6997 in which I found that there was almost no testing for the broadcasting behavior of methods defined to broadcast in `mtrand.pyx`.  Clarified documentation for these broadcastable functions with regards to output size depending on whether `size` is `None` and whether the inputs are all scalars.

Commit:
TST: Added broadcasting tests in test_random.py

Added a whole new suite of tests to ensure that
functions in mtrand.pyx which are broadcastable
actually broadcast their arguments properly.

jaimefrio commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -3426,7 +3426,8 @@ cdef class RandomState:
         samples : ndarray or scalar
-                  where the values are all integers in  [0, n].
+            Samples from logistic distribution, shaped according to
+            `size`. Otherwise, a single value is returned.
```
is the "Otherwise..." really needed? I feel that given the explanation on the `size` parameter, this is more confusing than helpful. Also, `size=None` does not necessarily mean a single return if the other parameters are arrays, right? Again, I'd make that change in `size`, not here.

gfyoung replied:
Indeed, that is true.  I think the documentation needs to be more standardized IMO, so I'll go through the doc strings for these broadcasting functions to make sure that that they properly explain how the size of the output varies depending on '''size'' and the size of the inputs.

gfyoung replied:
@jaimefrio : Docs all updated.

Commit:
DOC: Clarified output size for broadcastable mtrand.pyx functions

Clarified the output size depending on whether
scalar or non-scalar inputs are passed to
functions in mtrand.pyx that can broadcast
their arguments.

gfyoung commented:
Travis + Appveyor are happy, and the broadcasting + documentation (I think) have been hammered out pretty thoroughly now.  Ready to merge.

gfyoung commented:
Travis + Appveyor are happy, and the broadcasting + documentation (I think) have been hammered out pretty thoroughly now.  Ready to merge.

jaimefrio commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -1964,8 +1971,14 @@ cdef class RandomState:
-            single value is returned.
+            ``m * n * k`` samples are drawn.  If size is ``None`` (default),
+            a single value is returned if ``scale`` is a scalar.  Otherwise,
+            ``len(scale)`` samples are drawn.
```
This is not exactly correct: if `scale` is a 2D array, `len(scale)` is its number of rows, and this function returns `scale.size` items. If we wanted to write it in code, `np.array(scale).size` is probably the most accurate representation, but it may be better to come up with a description in prose, not sure...

gfyoung replied:
Done.  I prefer ```np.array(scale).size``` because it's more concise and should get the point across clear enough given the intended audience.  I had tried to describe these size-related values in more prose when working on this PR, but I couldn't come up with anything clear enough.

jaimefrio commented on numpy/random/mtrand/mtrand.pyx:
```diff
@@ -1468,13 +1468,14 @@ cdef class RandomState:
             Output shape.  If the given shape is, e.g., ``(m, n, k)``, then
-            ``m * n * k`` samples are drawn.  Default is None, in which case a
-            single value is returned.
+            ``m * n * k`` samples are drawn.  If size is ``None`` (default),
```
I guess we should modify the description of `low` to `float or array-like of floats, optional`, shouldn't we?

The same of course applies to all other broadcastable arguments.

gfyoung replied:
Fair enough.  Done.

jaimefrio commented:
Thanks for your patience with my finding incremental nitpicks, in it goes!

gfyoung commented:
No problem.  It's easier to fix a PR before it's merged than after. :smile:
